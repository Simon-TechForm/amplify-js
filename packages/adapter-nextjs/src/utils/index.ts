// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export { getAmplifyConfig } from './getAmplifyConfig';
export { createRunWithAmplifyServerContext } from './createRunWithAmplifyServerContext';
