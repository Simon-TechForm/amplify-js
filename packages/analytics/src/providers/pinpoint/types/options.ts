// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { PinpointServiceOptions } from '@aws-amplify/core/internals/providers/pinpoint';

/**
 * Options specific to Pinpoint identityUser.
 * Convert interface `IdentifyUserOptions` to a type to avoid `index signature 'string' is missing`
 * Taken from https://github.com/microsoft/TypeScript/issues/15300#issuecomment-1320421641
 */
export type IdentifyUserOptions = Pick<
	PinpointServiceOptions,
	keyof PinpointServiceOptions
>;
