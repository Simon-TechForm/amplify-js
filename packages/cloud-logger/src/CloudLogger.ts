/*
 * Copyright 2017-2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import { AWSS3Provider } from '@aws-amplify/storage';
import {
	ConsoleLogger,
	AWSCloudWatchProvider,
	Hub,
	Logger,
} from '@aws-amplify/core';
import { Auth } from '@aws-amplify/auth';
import CloudLoggerConnectivity from './cloudLoggerConnectivity';
import Cache from '@aws-amplify/cache';

const logger = new Logger('CloudLogger');
const CACHE_KEY_DEFAULT = 'cloud-logger/application-logs';

const DISABLED_EXPIRATION = 1000 * 60 * 60 * 12; // 12 hours

type RemoteConfigObject = {
	enabled: boolean;
	level: string;
	excludeClasses: string[];
	userAllowList: string[];
	refreshIntervalMinutes: number;
};

type RemoteConfig = {
	expires: number;
	config: RemoteConfigObject;
	hash: string;
};

type DisabledStatus = {
	expires: number;
};
class CloudLogger {
	private _config;
	private _storage: AWSS3Provider;
	private _remoteConfig: RemoteConfig;
	private _connectivity: CloudLoggerConnectivity;
	private _online: boolean = false;
	private _providerInstance: AWSCloudWatchProvider;
	private _cacheKey: string;
	private initialized = false;
	private user;
	private disabledStatus: DisabledStatus;
	private cache: typeof Cache;

	constructor() {
		this._preFlightCheck = this._preFlightCheck.bind(this);
	}

	async configure(conf) {
		logger.debug('configure', conf);
		if (!conf) {
			console.error('CloudLogger requires a config object');
		}

		this._config = conf;
		this._storage = new AWSS3Provider();
		this._storage.configure({
			Storage: {
				bucket: conf.Logging.configBucket,
				region: conf.aws_project_region,
				level: 'public',
			},
		});

		this._connectivity = new CloudLoggerConnectivity();
		this.cache = Cache.createInstance({
			keyPrefix: 'cl-prefix/',
			capacityInBytes: 2 * 1000 * 1000, // 2MB
			itemMaxSize: 500 * 1000, // 500KB
		});
	}

	private async initSignInListener() {
		logger.debug('initSignInListener');
		const sub = Hub.listen('auth', msg => {
			if (msg.payload?.event === 'signIn' && msg.payload?.data?.username) {
				logger.debug('initSignInListener - event');
				this.user = msg.payload.data;
				this._cacheKey = `${CACHE_KEY_DEFAULT}/${msg.payload.data.username}`;
				this.initSignOutListener();
				this.initialize();
				sub();
			}
		});
	}

	private async initSignOutListener() {
		logger.debug('initSignOutListener');
		const sub = Hub.listen('auth', msg => {
			if (msg.payload?.event === 'signOut') {
				logger.debug('initSignOutListener - event');
				this.stop();
				this.initSignInListener();
				sub();
			}
		});
	}

	private setDisabledFlag = async () => {
		const expires = Date.now() + DISABLED_EXPIRATION;

		const disabledStatus: DisabledStatus = {
			expires,
		};

		await this.cache.setItem(`${this._cacheKey}/disabled`, disabledStatus, {
			expires,
		});

		this.disabledStatus = disabledStatus;
	};

	private async checkAuth() {
		try {
			this.user = await Auth.currentAuthenticatedUser();
			logger.debug('authed user', this.user);
			this._cacheKey = `${CACHE_KEY_DEFAULT}/${this.user.username}`;
			this.initSignOutListener();

			if (!this.initialized) {
				this.initialize();
			} else {
				this.resume();
			}
		} catch (error) {
			if (error === 'The user is not authenticated') {
				this.initSignInListener();
			}
		}
	}

	async start() {
		this._start();
	}

	private validateLocalConfig() {
		if (!this._config.Logging) {
			throw new Error(
				'Must provide a `Logging` object in the config.\n For example: {\n  "Logging": {\n    "logGroupName": "my-log-group",\n    "configBucket": "my-s3-bucket",\n    "configFileName": "logging-config.json"\n  }\n}'
			);
		}

		const {
			Logging: { logGroupName, configBucket, configFileName },
			aws_project_region,
		} = this._config;

		let msg = '';

		if (logGroupName === undefined) {
			msg += 'Logging.logGroupName is required in the config\n';
		}
		if (configBucket === undefined) {
			msg += 'Logging.configBucket is required\n';
		}
		if (configFileName === undefined) {
			msg += 'Logging.configFileName is required\n';
		}
		if (aws_project_region === undefined) {
			msg += 'logGroupName is required\n';
		}

		if (msg) {
			throw new Error(
				`Invalid configuration for CloudLogger: ${msg}\n CloudLogger will not start;`
			);
		}
	}

	private async _start() {
		try {
			this.validateLocalConfig();

			this._connectivity.status().subscribe(async ({ online }) => {
				if (online && !this._online) {
					await this.checkAuth();
				} else if (!online) {
					this.pause();
				}
				this._online = online;
				logger.debug('network', online);
			});
		} catch (error) {
			logger.error(error);
		}
	}

	pause() {
		if (this._providerInstance) {
			this._providerInstance.pause();
		}
	}

	resume() {
		if (this._providerInstance) {
			this._providerInstance.resume();
		}
	}

	stop() {
		if (this._connectivity) {
			this._connectivity.unsubscribe();
		}
		this.cleanupRemoteLoggingResources();
		this.initialized = false;
	}

	clear() {
		this.clearCache();
	}

	private async clearCache() {
		this.cache.clear();
	}

	private async isDisabled() {
		const curTime = Date.now();

		const disabledInMem =
			this.disabledStatus != null && this.disabledStatus.expires > curTime;

		if (disabledInMem) {
			return true;
		}

		const checkCache =
			this.disabledStatus == null ||
			(this.disabledStatus != null && this.disabledStatus.expires <= curTime);

		if (checkCache) {
			const fromCache: DisabledStatus | undefined = await this.cache.getItem(
				`${this._cacheKey}/disabled`
			);

			if (fromCache != null && fromCache.expires > curTime) {
				this.disabledStatus = fromCache;
				return true;
			}
		}

		return false;
	}

	private async initialize() {
		await this._refreshConfig();
		if (!this.initialized) {
			this.initiateRemoteLogging();
		}
		this.initialized = true;
	}

	private async getConfig(): Promise<RemoteConfig> {
		logger.debug('Getting config from cache');
		const fromCache = await this.cache.getItem(this._cacheKey);

		if (fromCache) {
			logger.debug('Retrieved from cache');
			return fromCache;
		}

		// get config from s3
		try {
			logger.debug('Getting config from cloud');
			const configFileMeta = await this.getRemoteConfigMetadata();
			const remoteConfig = await this.getRemoteConfigObject();
			this.validateRemoteConfig(remoteConfig);
			const configObj = this.processRemoteConfig(
				remoteConfig,
				configFileMeta.eTag
			);
			logger.debug('Retrieved config from cloud');
			return configObj;
		} catch (error) {
			logger.debug('Error retrieving config from cloud', error);
			this.stop();
			throw error;
		}
	}

	private async _refreshConfig() {
		const configObj = await this.getConfig();

		await this.cache.setItem(this._cacheKey, configObj, {
			expires: configObj.expires,
		});
		this._remoteConfig = configObj;

		if (
			configObj.config.enabled === false ||
			configObj.config.enabled == null
		) {
			// Logging is disabled.
			logger.debug('remote config - logging is disabled globally');
			await this.setDisabledFlag();
		}

		const userAllowListDefined =
			configObj.config.userAllowList !== undefined &&
			configObj.config.userAllowList instanceof Array;

		if (userAllowListDefined) {
			const allowListEmpty = configObj.config.userAllowList.length === 0;
			const currentUserAllowListed = configObj.config.userAllowList.includes(
				this.user.username
			);

			if (allowListEmpty || !currentUserAllowListed) {
				logger.debug('remote config - logging is disabled for current user');
				await this.setDisabledFlag();
			}
		}
	}

	private async _preFlightCheck() {
		logger.debug('_preFlightCheck');

		if (await this.isDisabled()) {
			logger.debug('_preFlightCheck config disabled');
			return false;
		}

		if (this._remoteConfig && this._remoteConfig.expires) {
			if (this._remoteConfig.expires > Date.now()) {
				logger.debug('_preFlightCheck config valid');
				return true;
			}
		}

		logger.debug('_preFlightCheck in-memory config expired');
		await this.initialize();
		if (await this.isDisabled()) {
			logger.debug('_preFlightCheck post-refresh config disabled');
			return false;
		}
		return true;
	}

	private async getRemoteConfigMetadata() {
		const files = await this._storage.list('');
		const configFile = files.find(
			f => f.key === this._config.Logging.configFileName
		);

		if (!configFile) {
			throw new Error(
				'Remote config file not found in S3 bucket. CloudLogging is disabled'
			);
		}

		return configFile;
	}

	private async getRemoteConfigObject(): Promise<RemoteConfigObject> {
		try {
			const fileContents = await this._storage.get(
				this._config.Logging.configFileName,
				{
					download: true,
					cacheControl: 'no-cache',
				}
			);

			logger.debug('getRemoteConfigObject', fileContents);
			const body = <Blob>fileContents.Body;

			let jsonConfig;

			if (typeof body.text === 'function') {
				// body comes back as Blob on web
				jsonConfig = JSON.parse(await body.text());
			} else {
				// body comes back as Readable on RN
				jsonConfig = await new Response(body).json();
			}

			logger.debug('parsedConfig', jsonConfig);

			return jsonConfig;
		} catch (error) {
			logger.debug('Could not parse remote config', error);
			throw error;
		}
	}

	private validateRemoteConfig(jsonConfig: RemoteConfigObject) {
		if (jsonConfig.enabled === undefined) {
			throw new Error('remote config is missing `enabled: true`');
		}

		if (jsonConfig.refreshIntervalMinutes === undefined) {
			throw new Error(
				'remote config is missing `refreshIntervalMinutes` property'
			);
		}
	}

	private processRemoteConfig(
		jsonConfig: RemoteConfigObject,
		hash: string
	): RemoteConfig {
		logger.debug('json contents', jsonConfig);

		const expires = Date.now() + jsonConfig.refreshIntervalMinutes * 60 * 1000;

		const remoteConfigObj = {
			config: jsonConfig,
			hash,
			expires,
		};

		return remoteConfigObj;
	}

	private initiateRemoteLogging() {
		const providerConfig = {
			...this._config,
			Logging: { ...this._config.Logging, logStreamName: this._cacheKey },
		};

		this._providerInstance = new AWSCloudWatchProvider(providerConfig);
		this._providerInstance.setPreFlightCheck(this._preFlightCheck);

		ConsoleLogger.CLOUD_LOG_LEVEL =
			this._remoteConfig?.config?.level || 'DEBUG';
		ConsoleLogger.globalPluggables(this._providerInstance);
	}

	private cleanupRemoteLoggingResources() {
		ConsoleLogger.clearGlobalPluggables();

		if (this._providerInstance) {
			this._providerInstance.pause();
			this._providerInstance.clear();
			this._providerInstance = undefined;
		}
	}
}

const instance = new CloudLogger();
export { instance as CloudLogger };
