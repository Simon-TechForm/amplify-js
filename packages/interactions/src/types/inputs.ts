// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
	InteractionsMessage,
	InteractionsOnCompleteCallback,
} from './Interactions';

export type InteractionsSendInput = {
	botName: string;
	message: string | InteractionsMessage;
};

export type InteractionsOnCompleteInput = {
	botName: string;
	callback: InteractionsOnCompleteCallback;
};
