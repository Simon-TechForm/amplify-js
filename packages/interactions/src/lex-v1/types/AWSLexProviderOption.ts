// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { SendInput, OnCompleteInput } from './inputs';
import { SendOutput } from './outputs';

export interface AWSLexProviderOption {
	name: string;
	alias: string;
	region: string;
}

export interface IInteractions {
	send(input: SendInput): Promise<SendOutput>;
	onComplete(input: OnCompleteInput): void;
}
