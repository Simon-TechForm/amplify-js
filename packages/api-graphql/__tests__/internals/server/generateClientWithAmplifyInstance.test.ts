import * as raw from '../../../src';
import { Amplify, AmplifyClassV6, ResourcesConfig } from '@aws-amplify/core';
import { generateClientWithAmplifyInstance } from '../../../src/internals/server';
import configFixture from '../../fixtures/modeled/amplifyconfiguration';
import { Schema } from '../../fixtures/modeled/schema';
import { V6ClientSSRRequest, V6ClientSSRCookies } from '../../../src/types';

const serverManagedFields = {
	id: 'some-id',
	owner: 'wirejobviously',
	createdAt: new Date().toISOString(),
	updatedAt: new Date().toISOString(),
};

const config: ResourcesConfig = {
	API: {
		GraphQL: {
			apiKey: 'apikey',
			customEndpoint: undefined,
			customEndpointRegion: undefined,
			defaultAuthMode: 'apiKey',
			endpoint: 'https://0.0.0.0/graphql',
			region: 'us-east-1',
		},
	},
};

/**
 *
 * @param value Value to be returned. Will be `awaited`, and can
 * therefore be a simple JSON value or a `Promise`.
 * @returns
 */
function mockApiResponse(value: any) {
	return jest
		.spyOn((raw.GraphQLAPI as any)._api, 'post')
		.mockImplementation(async () => {
			const result = await value;
			return {
				body: {
					json: () => result,
				},
			};
		});
}

/**
 * For each call against the spy, assuming the spy is a `post()` spy,
 * replaces fields that are likely to change between calls (or library version revs)
 * with static values. When possible, on the unpredicable portions of these values
 * are replaced.
 *
 * ## THIS IS DESTRUCTIVE
 *
 * The original `spy.mocks.calls` will be updated *and* returned.
 *
 * For example,
 *
 * ```plain
 * headers.x-amz-user-agent: "aws-amplify/6.0.5 api/1 framework/0"
 * ```
 *
 * Is replaced with:
 *
 * ```plain
 * headers.x-amz-user-agent: "aws-amplify/latest api/latest framework/latest"
 * ```
 *
 * @param spy The Jest spy
 */
function normalizePostGraphqlCalls(spy: jest.SpyInstance<any, any>) {
	return spy.mock.calls.map((call: any) => {
		// The 1st param in `call` is an instance of `AmplifyClassV6`
		// The 2nd param in `call` is the actual `postOptions`
		const [_, postOptions] = call;
		const userAgent = postOptions?.options?.headers?.['x-amz-user-agent'];
		if (userAgent) {
			const staticUserAgent = userAgent.replace(/\/[\d.]+/g, '/latest');
			postOptions.options.headers['x-amz-user-agent'] = staticUserAgent;
		}
		// Calling of `post` API with an instance of `AmplifyClassV6` has been
		// unit tested in other test suites. To reduce the noise in the generated
		// snapshot, we hide the details of the instance here.
		return ['AmplifyClassV6', postOptions];
	});
}

describe('server generateClient', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	describe('with cookies', () => {
		test('subscriptions are disabled', () => {
			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			expect(() => {
				// @ts-expect-error
				client.models.Note.onCreate().subscribe();
			}).toThrow();
		});

		test('can list', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				filter: { name: { contains: 'name' } },
			});

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						headers: expect.objectContaining({
							'X-Api-Key': 'FAKE-KEY',
						}),
						body: {
							query: expect.stringContaining(
								'listTodos(filter: $filter, limit: $limit, nextToken: $nextToken)',
							),
							variables: {
								filter: {
									name: {
										contains: 'name',
									},
								},
							},
						},
					}),
				}),
			);

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						body: expect.objectContaining({
							// match nextToken in selection set
							query: expect.stringMatching(/^\s*nextToken\s*$/m),
						}),
					}),
				}),
			);

			expect(data.length).toBe(1);
			expect(data[0]).toEqual(
				expect.objectContaining({
					__typename: 'Todo',
					id: 'some-id',
					owner: 'wirejobviously',
					name: 'some name',
					description: 'something something',
				}),
			);
		});

		test('can list with nextToken', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				filter: { name: { contains: 'name' } },
				nextToken: 'some-token',
			});

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						headers: expect.objectContaining({
							'X-Api-Key': 'FAKE-KEY',
						}),
						body: {
							query: expect.stringContaining(
								'listTodos(filter: $filter, limit: $limit, nextToken: $nextToken)',
							),
							variables: {
								filter: {
									name: {
										contains: 'name',
									},
								},
								nextToken: 'some-token',
							},
						},
					}),
				}),
			);

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						body: expect.objectContaining({
							// match nextToken in selection set
							query: expect.stringMatching(/^\s*nextToken\s*$/m),
						}),
					}),
				}),
			);
		});

		test('can list with limit', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					listTodos: {
						items: [
							{
								__typename: 'Todo',
								...serverManagedFields,
								name: 'some name',
								description: 'something something',
							},
						],
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const { data } = await client.models.Todo.list({
				filter: { name: { contains: 'name' } },
				limit: 5,
			});

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						headers: expect.objectContaining({
							'X-Api-Key': 'FAKE-KEY',
						}),
						body: {
							query: expect.stringContaining(
								'listTodos(filter: $filter, limit: $limit, nextToken: $nextToken)',
							),
							variables: {
								filter: {
									name: {
										contains: 'name',
									},
								},
								limit: 5,
							},
						},
					}),
				}),
			);

			expect(spy).toHaveBeenCalledWith(
				expect.any(AmplifyClassV6),
				expect.objectContaining({
					options: expect.objectContaining({
						body: expect.objectContaining({
							// match nextToken in selection set
							query: expect.stringMatching(/^\s*nextToken\s*$/m),
						}),
					}),
				}),
			);
		});

		test('can custom query', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const spy = mockApiResponse({
				data: {
					echo: {
						resultContent: 'echo result content',
					},
				},
			});

			const getAmplify = async (fn: any) => await fn(Amplify);

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRCookies<Schema>
			>({
				amplify: getAmplify,
				config: config,
			});

			const result = await client.queries.echo({
				argumentContent: 'echo argumentContent value',
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
			expect(result?.data).toEqual({
				resultContent: 'echo result content',
			});
		});
	});
	describe('with request', () => {
		test('subscriptions are disabled', () => {
			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRRequest<Schema>
			>({
				amplify: null,
				config: config,
			});

			expect(() => {
				// @ts-expect-error
				client.models.Note.onCreate().subscribe();
			}).toThrow();
		});

		test('contextSpec param gets passed through to client.graphql', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRRequest<Schema>
			>({
				amplify: null,
				config: config,
			});

			const mockContextSpec = {};

			const spy = jest.spyOn(client, 'graphql').mockImplementation(async () => {
				const result: any = {};
				return result;
			});

			await client.models.Note.list(mockContextSpec);

			expect(spy).toHaveBeenCalledWith(
				mockContextSpec,
				expect.objectContaining({
					query: expect.stringContaining('listNotes'),
				}),
				{},
			);
		});

		test('can custom query', async () => {
			Amplify.configure(configFixture as any);
			const config = Amplify.getConfig();

			const client = generateClientWithAmplifyInstance<
				Schema,
				V6ClientSSRRequest<Schema>
			>({
				amplify: null,
				config: config,
			});

			const spy = jest.spyOn(client, 'graphql').mockImplementation(async () => {
				const result: any = {};
				return result;
			});

			const mockContextSpec = {};

			const result = await client.queries.echo(mockContextSpec, {
				argumentContent: 'echo argumentContent value',
			});

			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
		});
	});
});
