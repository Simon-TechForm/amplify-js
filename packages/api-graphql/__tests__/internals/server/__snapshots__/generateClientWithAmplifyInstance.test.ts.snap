// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server generateClient with cookies can custom query 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "abortController": AbortController {},
      "options": {
        "body": {
          "query": "query ($argumentContent: String!) {
  echo(argumentContent: $argumentContent) {
    resultContent
  }
}
",
          "variables": {
            "argumentContent": "echo argumentContent value",
          },
        },
        "headers": {
          "X-Api-Key": "FAKE-KEY",
          "x-amz-user-agent": "aws-amplify/latest api/latest framework/latest",
        },
        "signingServiceInfo": undefined,
        "withCredentials": undefined,
      },
      "url": "https://localhost/graphql",
    },
  ],
]
`;

exports[`server generateClient with request can custom query 1`] = `
[
  [
    "AmplifyClassV6",
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
		query($argumentContent: String!) {
			echo(argumentContent: $argumentContent) {resultContent}
		}
	",
      "variables": {
        "argumentContent": "echo argumentContent value",
      },
    },
  ],
]
`;
